### DISCUSSION TITLE
How do I get messages from the *agent* to the *groupchat manager* with request_reply?

### DISCUSSION THREAD

**Discussion by jspv** [microsoft/autogen#1050] [Q&A]
This is a simple LLM use case (no functions, no code execution).  I'm trying to get included on *every* message between agents via register_reply, but I'm only able to get the messages between the chat_manager to the agents, I never get messages from the agent back to the chat_manager (the messages that actually include the LLM response)

Looking at `run_chat` in `groupchat.py`, the messages from the agent to the groupchat manager are always sent by the     `speaker.send(reply, self, request_reply=False)` line, which of course will not trigger my reply function due to the `request_reply=False`.   Is there a way to get these messages?

 

```python
          try:
                # select the next speaker
                speaker = groupchat.select_speaker(speaker, self)
                # let the speaker speak
                reply = speaker.generate_reply(sender=self)
            except KeyboardInterrupt:
                # let the admin agent speak if interrupted
                if groupchat.admin_name in groupchat.agent_names:
                    # admin agent is one of the participants
                    speaker = groupchat.agent_by_name(groupchat.admin_name)
                    reply = speaker.generate_reply(sender=self)
                else:
                    # admin agent is not found in the participants
                    raise
            if reply is None:
                break
            # The speaker sends the message without requesting a reply
            speaker.send(reply, self, request_reply=False)
            message = self.last_message(speaker)
        return True, None
```

The following code will illustrate the problem by marking in the output whenever the registered reply function `get_agent_message` is called.

```python

from autogen import (
    Agent,
    AssistantAgent,
    UserProxyAgent,
    GroupChat,
    GroupChatManager,
)

config_list = **your stuff here**

llm_config = {"config_list": config_list, "cache_seed": 42, "stream": True}

assistant = AssistantAgent(
            "assistant", llm_config={"config_list": config_list}
        )

writer = AssistantAgent(
    name="writer",
    system_message=(
        "You are a writer of academic papers, you will be given a topic or"
        " prompt and you will provide an academic paper on that topic.  You"
        " work closely with an editor, when your editor rejects a paper, you"
        " will be given feedback on how to improve the paper and you will be"
        " expected to improve the paper and resubmit it to the editor."
    ),
    llm_config=llm_config,
)

editor = AssistantAgent(
    name="editor",
    system_message=(
        "You are a skilled editor of academic papers, has high standards and"
        " expects papers to be at a Masters level.  When papers are presented"
        " to you by a writer, you will either accept the paper or reject it. "
        " If you reject the paper, you will state immediately 'I reject this"
        " paper' and you then will provide specific feedback to the writer on"
        " how to improve the paper.  If you accept the paper, you will state"
        " immediately 'I accept this paper'.  If the writer later sumbits an"
        " updated version, review it again and provide feedback. To be"
        " accepted, the paper must have sufficient lenght, be free of"
        " grammatical errors, have referernces and citations, and demonstrate"
        " both creativity and a mastery of the subject matter.  Papers should"
        " be at least 4000 words, if papers are not long enough, provide"
        " suggestions on which paragraps and topics to expand the paper.  Be"
        " specific and explicit."
    ),
    llm_config=llm_config,
)

user_proxy = UserProxyAgent(
    name="User_proxy",
    system_message="A human admin.",
    human_input_mode="TERMINATE",
)

groupchat = GroupChat(
    agents=[user_proxy, writer, editor],
    messages=[],
    max_round=12,
)
manager = GroupChatManager(groupchat=groupchat, llm_config=llm_config)

def get_agent_message(recipient, messages, sender, config):
    message = messages[-1]["content"]
    print(">>>>>>>>>")
    print(f"got message from {sender.name} to {recipient.name} of length {len(message)}")
    if len(message) < 50:
        print(message)
    else:
        print(f"++ {message[:20]}..{message[-20:]}")
    print("<<<<<<<<<<")
    return False, None


user_proxy.register_reply(
    [Agent, None],
    reply_func=get_agent_message,
)
assistant.register_reply(
    [Agent, None],
    reply_func=get_agent_message,
)
writer.register_reply(
    [Agent, None],
    reply_func=get_agent_message,
)
editor.register_reply(
    [Agent, None],
    reply_func=get_agent_message,
)

user_proxy.initiate_chat(manager, clear_history=True,  message = "Tell me about the diet of the common house cat.")

```

**Comment by rickyloynd-microsoft** [comment]
@afourney

**Comment by thinkall** [comment]
Hi @jspv , group chat manager doesn't reply to any message, it just select a speaker and get reply from that speaker. If you want to process any message, register a function to the agents would be enough. Check #3279 .

**Reply by jspv** [reply]
Hi @thinkall, in my testing, registering a reply function to the agents does not get triggered with the agents are sending to the chat_manager, it is the recipient's reply function that is called, not the senders.  The code that sends the message from the agent to the chat manager is `speaker.send(reply, self, request_reply=False)` which specifically stops the reply function on the chat manager from being called.  

My goal is to capture each message to and from the agents.  My sample code registers a reply function on every Conversable agent that is intended to capture each communication.  

- Chat_manager -> (agent) Writer   (Writer reply function called, message captured)
- (agent) Writer -> chat_manager  (Chat_Manager reply function **not** called due to the `request_reply=False`, message not captured)
- chat_manager -> (agent) Editor (Editor reply function called, message captured)
- (agent) Editor -> chat_manager (Chat_Manager reply function **not** called due to the `request_reply=False`, message not captured)

How can I get the messages sent from the Agents back to the chat_manager?

**Reply by PersonaDev** [reply]
you could forward the output format from autogen into a log file and parse for info using the format, typically chat_manager does not communicate or take in replies, its mainly for backend managing the configs.

**Reply by thinkall** [reply]
> Hi @thinkall, in my testing, registering a reply function to the agents does not get triggered with the agents are sending to the chat_manager, it is the recipient's reply function that is called, not the senders. The code that sends the message from the agent to the chat manager is `speaker.send(reply, self, request_reply=False)` which specifically stops the reply function on the chat manager from being called.
> 
> My goal is to capture each message to and from the agents. My sample code registers a reply function on every Conversable agent that is intended to capture each communication.
> 
> * Chat_manager -> (agent) Writer   (Writer reply function called, message captured)
> * (agent) Writer -> chat_manager  (Chat_Manager reply function **not** called due to the `request_reply=False`, message not captured)
> * chat_manager -> (agent) Editor (Editor reply function called, message captured)
> * (agent) Editor -> chat_manager (Chat_Manager reply function **not** called due to the `request_reply=False`, message not captured)
> 
> How can I get the messages sent from the Agents back to the chat_manager?


Hi @jspv ,  the message  from  agent to chat_manager will be sent out to other agents directly. So, when  you see `chat_manager -> (agent)`, it's actually the `last_speaker -> current_speaker`.  With PR #3318 you'll be able to know the `last_speaker`.
